orchagent/fdborch.cpp

```diff
@@ -664,24 +1081,247 @@ bool FdbOrch::removeFdbEntry(const FdbEntry& entry)
     {
-        SWSS_LOG_ERROR("Failed to remove FDB entry. mac=%s, bv_id=0x%" PRIx64,
+        SWSS_LOG_ERROR("FdbOrch RemoveFDBEntry: Failed to remove FDB entry. mac=%s, bv_id=0x%lx",
                        entry.mac.to_string().c_str(), entry.bv_id);
         return true; //FIXME: it should be based on status. Some could be retried. some not
     }
 
+    SWSS_LOG_NOTICE("Removed mac=%s bv_id=0x%lx port:%s",
+            entry.mac.to_string().c_str(), entry.bv_id, port.m_alias.c_str());
+
+    port.m_fdb_count--;
+    m_portsOrch->setPort(port.m_alias, port);
+    vlan.m_fdb_count--;
+    m_portsOrch->setPort(vlan.m_alias, vlan);
     (void)m_entries.erase(entry);
 
+    // Remove in StateDb
+    if (fdbData.origin != FDB_ORIGIN_VXLAN_ADVERTIZED) 
+    {
+        m_fdbStateTable.del(key);
+    }
+
     gCrmOrch->decCrmResUsedCounter(CrmResourceType::CRM_FDB_ENTRY);
 
+    FdbUpdate update;
+    update.entry = entry;
+    update.port = port;
+    update.type = fdbData.type;
+    update.add = false;
+
+    notify(SUBJECT_TYPE_FDB_CHANGE, &update);
+
+    notifyTunnelOrch(update.port);
+
+    return true;
+}
+
+bool FdbOrch::flushFdbAll(bool flush_static)
+{
+    sai_status_t status;
+    sai_attribute_t port_attr;
+
+    if (!flush_static)
+    {
+        port_attr.id = SAI_FDB_FLUSH_ATTR_ENTRY_TYPE;
+        port_attr.value.s32 = SAI_FDB_FLUSH_ENTRY_TYPE_DYNAMIC;
+        status = sai_fdb_api->flush_fdb_entries(gSwitchId, 1, &port_attr);
+    }
+    else
+    {
+        status = sai_fdb_api->flush_fdb_entries(gSwitchId, 0, NULL);
+    }
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Flush fdb failed, return code %x", status);
+        return false;
+    }
+    return true;
+}
+
+bool FdbOrch::flushFdbByPort(const string &alias, bool flush_static)
+{
+    sai_status_t status;
     Port port;
-    m_portsOrch->getPortByBridgePortId(entry.bv_id, port);
+    sai_attribute_t port_attr[2];
 
-    SWSS_LOG_INFO("Notifying observers of FDB entry removal");
-    FdbUpdate update = {entry, port, false};
-    for (auto observer: m_observers)
+    if (!m_portsOrch->getPort(alias, port))
     {
-        observer->update(SUBJECT_TYPE_FDB_CHANGE, &update);
+        SWSS_LOG_ERROR("could not locate port from alias %s", alias.c_str());
+        return false;
     }
 
+    if ((port.m_bridge_port_id == SAI_NULL_OBJECT_ID) || !port.m_fdb_count)
+    {
+        /* port is not an L2 port or no macs to flush */
+        return true;
+    }
+
+    SWSS_LOG_NOTICE("m_bridge_port_id 0x%lx flush_static %d m_fdb_count %u", port.m_bridge_port_id, flush_static, port.m_fdb_count);
+
+    port_attr[0].id = SAI_FDB_FLUSH_ATTR_BRIDGE_PORT_ID;
+    port_attr[0].value.oid = port.m_bridge_port_id;
+    if (!flush_static)
+    {
+        port_attr[1].id = SAI_FDB_FLUSH_ATTR_ENTRY_TYPE;
+        port_attr[1].value.s32 = SAI_FDB_FLUSH_ENTRY_TYPE_DYNAMIC;
+        status = sai_fdb_api->flush_fdb_entries(gSwitchId, 2, port_attr);
```

Comment:
can we call this only once and use attr_count variable?

Reply:
Same comment applies to below two functions.

Reply:
Sure